File: remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // O(L) Time | O(1) Space
    // L : number of nodes in list
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy -> next = head;
        ListNode* currentNode = dummy;
        ListNode* nextNode = dummy;
        
        for (int i = 0; i <= n; ++i)
            nextNode = nextNode -> next;
        
        while (nextNode != nullptr) {
            currentNode = currentNode -> next;
            nextNode = nextNode -> next;
        }
        currentNode -> next = currentNode -> next -> next;
        return dummy -> next;
    }
};

File: Python/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        currentNode = dummy
        endNode = dummy

        for i in range(0, n + 1):
            endNode = endNode.next

        while endNode != None:
            currentNode = currentNode.next
            endNode = endNode.next

        currentNode.next = currentNode.next.next

        return dummy.next